New: adds UBL as additional output format for documents

alexmigf opened the pull request:
closes #481 

This PR adds support for UBL format in companion with the existing PDF output. This will replace our UBL addon, and future developments on UBL must be done on this plugin instead.

### Specs

* Code migrated from the UBL addon to this plugin
* Load `sabre\xml` dependency and updated to version `2.2.5` (supports PHP 7.1+, min version required for this plugin)
* Detects if the UBL addon is installed and disables it. Also adds a notice stating that and suggesting the addon deletion
* Migrates UBL addon settings to this plugin settings on update to the future `3.6.0` version
* Shows raw `xml` output on preview when UBL tab is selected in the document settings
* Only Invoice has UBL support by default, other documents can also have it using this hook: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/521/files#diff-b4b45c08877ccc6f215e534426a81b3edd82ce371cfc3fce7420e62e279b3c1eR122

### Screenshots

Document settings for UBL format:

![Screenshot from 2023-06-02 14-17-37](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/533273/8433c9aa-d399-45bb-bb34-79dc03aa01e2)

UBL output preview:

![Screenshot from 2023-06-02 14-21-45](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/533273/1f107b22-b5f7-4a3b-a5fe-42f6aa58d4ab)


Taxes classification:

![Screenshot from 2023-06-02 14-18-10](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/533273/006c2c59-f078-456a-9261-bba0417d7eb2)


New dynamic action button (can be generated for other UBL documents):

![Screenshot from 2023-06-02 14-18-37](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/533273/18e282fa-ca02-4c54-af2c-1994bb615c09)


Bulk actions support:

![Screenshot from 2023-06-02 14-19-37](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/533273/c499edfd-2ab4-465b-bf92-b7994bc1e3e9)

New independent meta box for UBL in the order page:

![Screenshot from 2023-06-02 14-20-04](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/533273/657ef330-1cf4-4198-aa8a-c489b8c94dd3)

UBL document settings export/import support:

![Screenshot from 2023-06-05 11-59-49](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/533273/d2bad21c-1036-4995-a5da-872f7a7abbda)

dwalkerpriv reviewed (changes requested):
Some of my initial tests:

- The admin notice about the deactivation works.
- UI seems stable
- Tax classes load with the correct values
- XML file attaches normally to emails and opens normally after download

Some things I've noticed:
- Disabling the admin notice while on the plugin page page redirects the PDF Invoices page. Not sure if this is the same for all pages
- UBL plugin is not deactivated if reactivated. This deactivates the built it UBL feature. Deactivating the UBL plugin brings back the built in UBL features
- UBL invoice creation button on order edit page isn't there
![Screenshot 2023-06-15 170615](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/62675729/1ddffa06-4edf-4719-b666-761a51965b3f)

- UBL icon does not have the green tick when the ubl file is created
![Screenshot 2023-06-15 171126](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/62675729/5db0fc7c-71a6-4116-82ed-6b49f26772d8)

- Clicking the UBL button on the Orders page creates and opens a PDF in the browser with the HTML option enable in the Status page.
- Clicking the UBL button on the Orders page with the HTML option disabled creates the XML file. The files opens fine after download

alexmigf commented:
Thanks @dwalkerpriv 

> Disabling the admin notice while on the plugin page page redirects the PDF Invoices page. Not sure if this is the same for all pages

A: Yes, will always redirect to the plugin settings page.

> UBL plugin is not deactivated if reactivated. This deactivates the built it UBL feature. Deactivating the UBL plugin brings back the built in UBL features

A: Not sure I understood. Can you explain better?

> UBL invoice creation button on order edit page isn't there

A: You have a new metabox just for UBL documents, see here: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/521#issue-1738193295

> UBL icon does not have the green tick when the ubl file is created
A: UBL never had the green tick, it's an instant document, is always available. Of course, if you didn't created the PDF Invoice yet when you click the UBL button it will force the PDF Invoice to be generated.

> Clicking the UBL button on the Orders page creates and opens a PDF in the browser with the HTML option enable in the Status page.

A: Not for me. Maybe you need to clear your browser cache.

dwalkerpriv commented:
> UBL plugin is not deactivated if reactivated. This deactivates the built it UBL feature. Deactivating the UBL plugin brings back the built in UBL features

@alexmigf  So when you install the update, the UBL add-on plugin is auto disabled. If you were to reactivate the UBL add-on plugin again, the built in UBL features from this is deactivated and the UBL add-on plugin is used instead.

Deactivating the UBL add-on plugin brings back the built in UBL features from this update.

Perhaps there should be a persisant check for the activation of the add-on plugin and then disable it so the built in UBL features from this update are always used?

alexmigf commented:
@dwalkerpriv yes, we can do that, but I was expecting it to be deactivated again when you try to activate it again, but I will check that.

MohamadNateqi reviewed (changes requested):
I have completed a portion of the review.
Next, I will review the UBL directory files and also conduct a review for WPCS.

alexmigf commented on includes/documents/class-wcpdf-bulk-document.php:
```diff
@@ -41,6 +41,9 @@ public function __construct( $document_type, $order_ids = array() ) {
 		$this->is_bulk   = true;
+		
+		// output formats (placed after parent construct to override the abstract default)
+		$this->output_formats = apply_filters( "wpo_wcpdf_{$this->type}_output_formats", [ 'pdf' ], $this );
```
I just wanted to have consistency between all documents, even understanding that this is a different kind of document. Now because I have added the `is_enabled()` method, I'm using this property.

alexmigf commented on includes/documents/class-wcpdf-invoice.php:
```diff
@@ -186,13 +192,42 @@ public function get_filename( $context = 'download', $args = array() ) {
+					break;
+				case 'ubl':
+					$page = $option_group = $option_name = "wpo_wcpdf_documents_settings_{$this->get_type()}_{$output_format}";
+					$settings_fields = $this->get_ubl_settings_fields( $option_name );
```
No, you have one below, and if you look above the PDF one is legacy.

alexmigf commented on includes/class-wcpdf-settings-ubl.php:
```diff
@@ -0,0 +1,145 @@
+		</div>
+		<?php
+		
+		switch ( $active_section ) {
```
I know, is just prepare to be expanded. You can have different sections here, of course at this moment we just have 1 because I have removed the other and placed in the document settings.

alexmigf commented on includes/class-wcpdf-settings-documents.php:
```diff
@@ -50,10 +57,36 @@ public function output( $section ) {
+								if ( $document_output_format == 'ubl' ) {
+									$tab_title .= ' <sup class="wcpdf_beta">beta</sup>';
+								}
+								printf( '<a href="%1$s" class="nav-tab nav-tab-%2$s %3$s">%4$s</a>', esc_url( add_query_arg( 'output_format', $document_output_format ) ), esc_attr( $document_output_format ), $active, $tab_title );
```
In the menu is also clickable, I think we can keep the same way here.

alexmigf commented on includes/class-wcpdf-documents.php:
```diff
@@ -65,19 +65,31 @@ public function init() {
 	 */
-	public function get_documents( $filter = 'enabled' ) {
-		if ( empty($this->documents) ) {
+	public function get_documents( $filter = 'enabled', $output_format = 'pdf' ) {
```
The problem is that we are using this function in the extensions, and that requires some compatibility check. I think this PR is a bit complex, maybe we should do that in another PR.

alexmigf commented on includes/documents/abstract-wcpdf-order-document.php:
```diff
@@ -757,6 +794,26 @@ public function shop_name() {
+	public function get_shop_vat_number() {
+		return $this->get_settings_text( 'vat_number', '', false );
+	}
+	public function shop_vat_number() {
```
I thought about that some time ago, I think I discuss that with Ewout, the previous owner of the company, and his idea was to avoid having the `echo` in the PDF templates to simplify the code there. I think we should keep like that for now.

YordanSoares reviewed (commented):
## Steps followed to test this PR:

1. Installed the UBL extension
2. Configured the UBL general settings:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/30989a3b-2027-4cc0-9765-d650b7fb160b)
3. Configured the UBL taxes settings:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/124bc1d8-c944-4802-8e71-384744a19348)

4. Installed the branch from this PR
5. The UBL extension was disabled and the warning notice was displayed as expected:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/575f4256-9c3a-494e-b1cb-1c827b3ec338)
6. The "Attach to" and "Include encrypted PDF" settings were imported correctly:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/2222909f-e4ab-4f26-a98c-5fe9d3c8121d)
7. The same happened with the shop VAT and Chamber of Commerce numbers:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/e2c9e83b-0e4b-4326-83e7-f520c5cbc0dd)
   I would expect that the company name were imported as the shop name, but this didn't happen. However, I'm aware that the shop name could have being set previously, but wanted to mention this, just in case.

8. Also, tax settings were imported correctly:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/038b6b4e-ff32-4165-a53f-24223f21c037)

9. Generated the UBL invoice, and it was downloaded correctly, but ignoring the encrypted PDF, although the option to do it was activated.

## Remarks:

1. As mentioned above, in the step 7, the company name was not imported as the shop name: Is this the expected behavior?
2. The UBL invoice seem to be ignoring the encrypted PDF setting, unless you enable the "Test mode". If you disable the text mode, the UBL invoices again ignores the setting, if enabled.
3. As @dwalkerpriv already mentioned, you can re-enable the UBL extension, then both settings will be displayed, included those from the UBL tab: 

   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/8d06e865-5e2d-437c-941c-ce135c02f2a7)
 
   I think we should avoid enabling this setting from the next public release on.
4. The UBL settings exported only includes the settings under **WooCommerce > PDF Invoices > Documents > Invoice > UBL**, but the shop VAT and CoC numbers, and the tax settings are not included:
   
   ```
   {"type":"invoice_ubl","settings":{"enabled":"1","attach_to_email_ids":{"customer_processing_order":"1","customer_completed_order":"1"},"include_encrypted_pdf":"1"}}```
5. The UBL settings reset is not yet supported.
6. The bulk export for the UBL invoice seems not to be working. When I select several orders and try to export their UBL invoice, it's only exported the one from the last order of the selected. 
7. Although I would leave the order list icon untouched (I like it!), what do you think about adding a green checkmark to the UBL button too, within the order details, when the PDF invoice is created?:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/57811729-9df7-4074-ad11-c6be5c21018b)

alexmigf commented on assets/js/order-script.js:
```diff
@@ -1,34 +1,55 @@
+			if ( ubl_output ) {
+				$.each( checked, function( i, order_id ) {
+					full_url = partial_url + '&order_ids='+order_id+'&output=ubl';
+					window.open( full_url, '_blank' );
```
I followed your suggestion but I forgot that this was inside a `each` loop. I have reverted this.

MohamadNateqi replied:
Oh okay, sorry for the wrong suggestion.

alexmigf commented:
@YordanSoares 

1. Fixed, but migrates only if the `shop_name`is empty.
2. Fixed!
3. Fixed!
4. The VAT and COC fields are inside the General Settings now. The UBL Tax Settings are now available in the export too.
5. Fixed!
6. Fixed!
7. Done!

YordanSoares reviewed (changes requested):
@alexmigf most of the issues are fixed now.

However, there are still a few additional tweaks to be made:

1. The UBL Taxes and UBL Invoice settings reset seems not to be working, yet:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/6137b5c2-691b-4106-9edc-998ac0b50365)
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/efbd54fb-26e0-47be-97fa-26762b5b480f)

2. On the other hand, the import for the UBL Taxes settings is not working either:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/b6f2bce0-9ce5-4dc0-9f97-c3d744521cd8)


3. I think that displaying the same icon for both the PDF and the UBL invoices when the document is not yet created may be confusing:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/7098af53-4442-4a7d-aed2-1242e2e6a279)
   
   I would always let the "code syntax" sign for the UBL one, even if the invoice is not created yet, how it was working previously:
   
   ![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/bbcf950d-a63f-41ee-9f1f-f8b3c297eaa8)

alexmigf commented:
@YordanSoares fixed!

alexmigf commented on includes/documents/abstract-wcpdf-order-document.php:
```diff
@@ -915,6 +965,43 @@ public function output_html() {
+
+		ob_clean();
+		flush();
+		readfile( $full_filename );
```
Can you explain your point of view?

MohamadNateqi replied:
I was wondering if using `@readfile()` instead of `readfile()` is a better option to hide error messages, as done in `wcpdf-pro-bulk-export.php`.

alexmigf commented on includes/class-wcpdf-documents.php:
```diff
@@ -55,24 +55,48 @@ public function init() {
+						}
+						break;
+					default:
+						foreach ( $document->output_formats as $document_output_format ) {
```
https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/04fc4aa10d68384c12ae4f0297f179aa4ac60082/includes/class-wcpdf-documents.php#L60

Basically when you pass any other string it should correspond to any.

MohamadNateqi replied:
In that case, will it be possible to have a situation where this function is being used to get documents that are enabled for a new output format created by a developer?

alexmigf replied:
Yes, doesn't hurt supporting that. I will do that.

YordanSoares reviewed (commented):
@alexmigf hidding the UBL invoice if the PDF invoice is not created is even better 👍 

![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/76b246cd-ca36-47aa-bf34-5216dfbf4217)

Regarding the reset settings tool, I still have this error for the UBL invoice reset:

![image](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/assets/38109855/fafca2ac-86ec-4c17-98d2-a6f2c8f816bc)

Maybe it is missing the `invoice_ubl` key?

alexmigf commented:
@YordanSoares fixed!

YordanSoares reviewed (approved):
Tested exporting, reset and exporting for the UBL Invoice settings and LGTM!

MohamadNateqi commented:
@alexmigf Sure. I'll soon.

MohamadNateqi commented on ubl/Handlers/Ubl/PaymentMeansHandler.php:
```diff
@@ -0,0 +1,24 @@
+class PaymentMeansHandler extends UblHandler {
+	
+	public function handle( $data, $options = array() ) {
+		$paymentMeans = arraY(
```
WPCS

MohamadNateqi commented on ubl/Collections/Collection.php:
```diff
@@ -0,0 +1,41 @@
+abstract class Collection implements Iterator {
+	
+	/** @var array */
+	private $items = [];
```
WPCS

MohamadNateqi commented on ubl/Documents/Document.php:
```diff
@@ -0,0 +1,191 @@
+	abstract public function getData();
+
+	public function getTaxRates() {
+		$order_tax_data = [];
```
WPCS

MohamadNateqi commented on ubl/Documents/UblDocument.php:
```diff
@@ -0,0 +1,91 @@
+
+		foreach ( $this->getFormat() as $key => $value ) {
+			$handler = new $value['handler']($this);
+			$options = isset( $value['options'] ) && is_array( $value['options'] ) ? $value['options'] : [];
```
WPCS - short array syntax

MohamadNateqi commented on ubl/Settings/TaxesSettings.php:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	public function getAvailableSchemes() {
+		return [
```
WPCS - array

MohamadNateqi commented on ubl/Settings/TaxesSettings.php:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	public function getAvailableCategories() {
+		return [
```
WPCS - array

MohamadNateqi commented on includes/documents/class-wcpdf-bulk-document.php:
```diff
@@ -32,11 +32,39 @@ class Bulk_Document {
 		$this->is_bulk   = true;
+		
+		// output formats (placed after parent construct to override the abstract default)
+		$this->output_formats = apply_filters( "wpo_wcpdf_{$this->type}_output_formats", [ 'pdf' ], $this );
```
WPCS - array

MohamadNateqi commented on includes/documents/class-wcpdf-invoice.php:
```diff
@@ -42,8 +43,11 @@ public function __construct( $order = 0 ) {
 		parent::__construct( $order );
+		
+		// output formats (placed after parent construct to override the abstract default)
+		$this->output_formats = apply_filters( "wpo_wcpdf_{$this->type}_output_formats", [ 'pdf', 'ubl' ], $this );
```
WPCS - array

MohamadNateqi commented on includes/documents/class-wcpdf-invoice.php:
```diff
@@ -181,13 +187,41 @@ public function get_filename( $context = 'download', $args = array() ) {
+		do_action( "wpo_wcpdf_before_{$this->type}_init_settings", $this );
+		
+		foreach ( $this->output_formats as $output_format ) {
+			$settings_fields = [];
```
WPCS - array

MohamadNateqi commented on includes/documents/class-wcpdf-packing-slip.php:
```diff
@@ -23,9 +23,12 @@ public function __construct( $order = 0 ) {
 		parent::__construct( $order );
+		
+		// output formats (placed after parent construct to override the abstract default)
+		$this->output_formats = apply_filters( "wpo_wcpdf_{$this->type}_output_formats", [ 'pdf' ], $this );
```
WPCS - array

MohamadNateqi commented on includes/settings/class-wcpdf-settings-debug.php:
```diff
@@ -318,13 +325,17 @@ public function import_settings( $data ) {
 		}
 		
-		if ( in_array( $type, [ 'general', 'debug' ] ) ) {
+		if ( in_array( $type, [ 'general', 'debug', 'ubl_taxes' ] ) ) {
```
WPCS - array

MohamadNateqi commented on includes/settings/class-wcpdf-settings-debug.php:
```diff
@@ -448,7 +467,14 @@ public function get_setting_types() {
+			
+			foreach ( $document->output_formats as $output_format ) {
+				$slug = $document->get_type();
+				if ( $output_format != 'pdf' ) {
```
WPCS

MohamadNateqi commented on includes/settings/class-wcpdf-settings-documents.php:
```diff
@@ -59,10 +70,36 @@ public function output( $section ) {
+					<h2 class="nav-tab-wrapper">
+						<?php
+							foreach ( $section_document->output_formats as $document_output_format ) {
+								$active    = ( $output_format == $document_output_format ) || ( $output_format != 'pdf' && ! in_array( $output_format, $section_document->output_formats ) ) ? 'nav-tab-active' : '';
```
WPCS - loose comparison

MohamadNateqi commented on includes/settings/class-wcpdf-settings-documents.php:
```diff
@@ -59,10 +70,36 @@ public function output( $section ) {
+							foreach ( $section_document->output_formats as $document_output_format ) {
+								$active    = ( $output_format == $document_output_format ) || ( $output_format != 'pdf' && ! in_array( $output_format, $section_document->output_formats ) ) ? 'nav-tab-active' : '';
+								$tab_title = strtoupper( esc_html( $document_output_format ) );
+								if ( $document_output_format == 'ubl' ) {
```
WPCS - loose comparison

MohamadNateqi commented on includes/settings/class-wcpdf-settings-documents.php:
```diff
@@ -59,10 +70,36 @@ public function output( $section ) {
-		do_settings_sections( "wpo_wcpdf_documents_settings_{$section}" );
-		submit_button();
+			$output_format_compatible = false;
+			if ( $output_format != 'pdf' && in_array( $output_format, $section_document->output_formats ) ) {
```
WPCS - loose comparison

MohamadNateqi commented on includes/settings/class-wcpdf-settings-documents.php:
```diff
@@ -59,10 +70,36 @@ public function output( $section ) {
+				$output_format_compatible = true;
+			}
+			
+			$option_name = ( $output_format == 'pdf' || ! $output_format_compatible ) ? "wpo_wcpdf_documents_settings_{$section}" : "wpo_wcpdf_documents_settings_{$section}_{$output_format}";
```
WPCS - loose comparison

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -202,38 +202,76 @@ public function update_pdf_counter( $document_type, $document ) {
+								$document_title   = is_callable( array( $document, 'get_title' ) ) ? $document->get_title() : $document_title;
+								$document_exists  = is_callable( array( $document, 'exists' ) ) ? $document->exists() : false;
+								$document_printed = $document_exists && is_callable( array( $document, 'printed' ) ) ? $document->printed() : false;
+								$class            = [ $document->get_type(), $output_format ];
```
WPCS - array

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -202,38 +202,76 @@ public function update_pdf_counter( $document_type, $document ) {
+							break;
+						case 'ubl':
+							if ( $document->is_enabled( $output_format ) ) {
+								$document_url    = WPO_WCPDF()->endpoint->get_document_link( $order, $document->get_type(), [ 'output' => $output_format ] );
```
WPCS - array

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -202,38 +202,76 @@ public function update_pdf_counter( $document_type, $document ) {
+								$document_url    = WPO_WCPDF()->endpoint->get_document_link( $order, $document->get_type(), [ 'output' => $output_format ] );
+								$document_title  = is_callable( array( $document, 'get_title' ) ) ? $document->get_title() : $document_title;
+								$document_exists = is_callable( array( $document, 'exists' ) ) ? $document->exists() : false;
+								$class           = [ $document->get_type(), $output_format ];
```
WPCS - array

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -202,38 +202,76 @@ public function update_pdf_counter( $document_type, $document ) {
+									$class[] = 'exists';
+								}
+								
+								$listing_actions[$document->get_type()."_{$output_format}"] = array(
```
WPCS

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -244,18 +282,23 @@ public function add_listing_actions( $order ) {
-			);
+			
+			// ubl replaces exists
+			$exists  = isset( $data['output_format'] ) && $data['output_format'] == 'ubl' ? '<svg class="icon-ubl" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M8.59323 18.3608L9.95263 16.9123L9.95212 16.8932L4.85783 12.112L9.64826 7.00791L8.18994 5.63922L2.03082 12.2016L8.59323 18.3608ZM15.4068 18.3608L14.0474 16.9123L14.0479 16.8932L19.1422 12.1 ...
```
WPCS - loose comparison

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -563,6 +620,81 @@ public function pdf_actions_meta_box( $post_or_order_object ) {
+				$document       = wcpdf_get_document( $document->get_type(), $order );
+				
+				if ( $document ) {
+					$document_url    = WPO_WCPDF()->endpoint->get_document_link( $order, $document->get_type(), [ 'output' => 'ubl' ] );
```
WPCS - array

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -563,6 +620,81 @@ public function pdf_actions_meta_box( $post_or_order_object ) {
+					$document_url    = WPO_WCPDF()->endpoint->get_document_link( $order, $document->get_type(), [ 'output' => 'ubl' ] );
+					$document_title  = is_callable( array( $document, 'get_title' ) ) ? $document->get_title() : $document_title;
+					$document_exists = is_callable( array( $document, 'exists' ) ) ? $document->exists() : false;
+					$class           = [ $document->get_type(), 'ubl' ];
```
WPCS - array

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -563,6 +620,81 @@ public function pdf_actions_meta_box( $post_or_order_object ) {
+						$class[] = 'exists';
+					}
+					
+					$meta_box_actions[$document->get_type()] = array(
```
WPCS

MohamadNateqi commented on includes/class-wcpdf-admin.php:
```diff
@@ -563,6 +620,81 @@ public function pdf_actions_meta_box( $post_or_order_object ) {
+					}
+				}
+				
+				if ( $ubl_documents == 0 ) {
```
WPCS

MohamadNateqi commented on includes/wcpdf-functions.php:
```diff
@@ -109,30 +110,72 @@ function wcpdf_get_packing_slip( $order, $init = false ) {
 }
 
+function wcpdf_get_bulk_actions() {
+	$actions   = [];
```
WPCS - array

MohamadNateqi commented on includes/wcpdf-functions.php:
```diff
@@ -109,30 +110,72 @@ function wcpdf_get_packing_slip( $order, $init = false ) {
+	foreach ( $documents as $document ) {
+		foreach ( $document->output_formats as $output_format ) {
+			$slug = $document->get_type();
+			if ( $output_format != 'pdf' ) {
```
WPCS - loose comparison

MohamadNateqi reviewed (approved):
Since you have applied WPCS on the UBL directory, it would be great if you could rename the function to follow the sneak case, or we can create another issue as the changes may be significant.

MohamadNateqi reviewed (approved):
Great work, Alex!
Thank you for your efforts.

dwalkerpriv reviewed (approved):
Works fine on my end. Settings exported and imported without issues as well

alexmigf commented:
I was thinking that maybe we should rename the plugin following this implementation. Joe have mentioned this in past meetings, maybe something like this: 

* PDF & UBL Invoices & Packing Slips for WooCommerce

But I don't like seeing two `&` in the title, maybe we should apply a `/` or a `,`?

What do you guys think?

@dwalkerpriv @YordanSoares @MohamadNateqi @Terminator-Barbapapa

MohamadNateqi commented:
@alexmigf I agree with you that having two "&" symbols can look cluttered.
I think "PDF/UBL Invoices & Packing Slips for WooCommerce" seems good.

Also, I wanted to say that I like the idea of renaming, as it improves clarity and reflects this new feature that can also be good in terms of SEO, but I'm also sort of worried about the impact of this change as this plugin has been around for years with this name.

alexmigf commented:
> but I'm also sort of worried about the impact of this change as this plugin has been around for years with this name.

Shouldn't have a huge impact, we have also renamed it recently to change the "for WooCommerce" to the end of the name.

I think we should have some space outside the `/` because it terms of SEO this reads like this:

* PDF/UBL => PDFUBL
* PDF / UBL => PDF UBL

The second seems better to me.

MohamadNateqi commented:
Yeah, having some space around `/` is better.

YordanSoares commented:
This makes me recall that in Spain people often joke about how good is our plugin, but how difficult is to remember its name.

To be honest, if we're willing to rename the plugin, I would prefer to run a brainstorming and look for a branding short name, not a name that states-all-the-document-types-our-plugin-offers.

Then we can mention what our plugin does as a subtitle or within the plugin descriptions in both on WordPress.org and our store. For instance:

- On WordPress.org: 
   - NewPluginName - PDF / UBL Invoices & Packing Slips for WooCommerce
- On our store, the new extension names would be shorted and easy to remember and differentiate (our customer often buy a different plugin because I think they get mess with the long names):
   - NewPluginName - Professional
   - NewPluginName - Premium Templates

Here, some naming ideas to start with:
- EasyInvoicing
- QuickInvoice
- FastInvoicing
- InvoicingWise
- WooInvoicing

alexmigf commented:
By rename I didn't mean change all name, just a way to include the UBL there somehow. Full rename, or brand rename should be discussed privately in Slack I think, and maybe including the marketing department from UpdraftPlus.

Terminator-Barbapapa commented on includes/settings/class-wcpdf-settings-general.php:
```diff
@@ -170,6 +170,32 @@ public function init_settings() {
 					'translatable'	=> true,
 				)
 			),
+			array(
+				'type'     => 'setting',
+				'id'       => 'vat_number',
+				'title'    => __( 'Shop VAT Number', 'woocommerce-pdf-invoices-packing-slips' ),
+				'callback' => 'text_input',
+				'section'  => 'general_settings',
+				'args'     => array(
+					'option_name'  => $option_name,
+					'id'           => 'vat_number',
+					'translatable' => true,
+					'description'  => __( 'Required by the UBL format.', 'woocommerce-pdf-invoices-packing-slips' ),
+				)
+			),
+			array(
+				'type'     => 'setting',
+				'id'       => 'coc_number',
+				'title'    => __( 'Shop Chamber of Commerce Number', 'woocommerce-pdf-invoices-packing-slips' ),
+				'callback' => 'text_input',
+				'section'  => 'general_settings',
+				'args'     => array(
+					'option_name'  => $option_name,
+					'id'           => 'coc_number',
+					'translatable' => true,
+					'description'  => __( 'Required by the UBL format.', 'woocommerce-pdf-invoices-packing-slips' ),
+				)
+			),
```
I think these settings should only display in 'General' when the UBL document is enabled. Also I would change the description to: `Required for UBL file generation.`

alexmigf replied:
I wouldn't do that, because the code is prepared to have UBL for other documents. If we go that way we should also disable the UBL tab too, which I don't think we should do.

Of course it's doable to hide them, but I see this setting like the `Extra field 1`, `Extra field 2` and `Extra field 3`, which are being displayed even when the Premium Templates is not installed. Beside that, I'm seeing customers using the VAT number field to display in the PDF as well, which can be also added there easily.

Terminator-Barbapapa replied:
I would propose to remove the `Required by the UBL format.` description from these settings. And add an admin notice when a UBL document is enabled but certain required settings have no value yet. Something like the following:

`You've enabled UBL format for invoices, but some essential details are missing. Please ensure you've added your VAT and CoC numbers in the General settings [link]. Also, specify your tax rates in the UBL settings [link].`

Also I think we need to create a documentation page for setting up the UBL format. We can link to it from the UBL tab and also from the UBL settings in the Documents tab.

alexmigf replied:
I agree, seems a good way to handle it.

@YordanSoares can you create the issue/doc page for the UBL?

Terminator-Barbapapa commented on woocommerce-pdf-invoices-packingslips.php:
```diff
@@ -467,6 +475,73 @@ public function rtl_detected() {
+			ob_start();
+			?>
+			<div class="notice notice-warning">
+				<p><?php _e( 'The UBL addon was installed in your store but is no longer needed. We have disable it when you upgraded the PDF Invoices & Packing Slips for WooCommerce plugin. You are free to uninstall it.', 'woocommerce-pdf-invoices-packing-slips' ); ?></p>
```
I would suggest changing this to: `While updating the ‘PDF Invoices & Packing Slips for WooCommerce’ plugin we’ve noticed our UBL add-on was active on your site. This functionality is now incorporated into the core plugin. We’ve deactivated the add-on for you, and you are free to uninstall it.`

Terminator-Barbapapa commented on ubl/Documents/UblDocument.php:
```diff
@@ -0,0 +1,115 @@
+class UblDocument extends Document {
+	
+	public function get_format() {
+		$format = apply_filters( 'wpo_wc_ubl_document_fornat' , array(
```
Spelling mistake in the filter name: `fornat` should be `format`.

Terminator-Barbapapa commented on includes/settings/class-wcpdf-settings-ubl.php:
```diff
@@ -139,6 +142,21 @@ public function save_order_taxes( $order ) {
+			$message = sprintf(
+				/* translators: 1. General Settings, 2. UBL Tab  */
+				__( 'You\'ve enabled UBL output for a document, but some essential details are missing. Please ensure you\'ve added your VAT and CoC numbers in the %1$s. Also, specify your tax rates in the %2$s.', 'woocommerce-pdf-invoices-packing-slips' ),
+				'<a href="' . esc_url( admin_url( 'admin.php?page=wpo_wcpdf_options_page' ) ) . '" target="_blank">' . __( 'General Settings', 'woocommerce-pdf-invoices-packing-slips' ) . '</a>',
```
I would make it `General settings` (settings lowercase).

Terminator-Barbapapa commented on includes/settings/class-wcpdf-settings-ubl.php:
```diff
@@ -139,6 +142,21 @@ public function save_order_taxes( $order ) {
+				/* translators: 1. General Settings, 2. UBL Tab  */
+				__( 'You\'ve enabled UBL output for a document, but some essential details are missing. Please ensure you\'ve added your VAT and CoC numbers in the %1$s. Also, specify your tax rates in the %2$s.', 'woocommerce-pdf-invoices-packing-slips' ),
+				'<a href="' . esc_url( admin_url( 'admin.php?page=wpo_wcpdf_options_page' ) ) . '" target="_blank">' . __( 'General Settings', 'woocommerce-pdf-invoices-packing-slips' ) . '</a>',
+				'<a href="' . esc_url( admin_url( 'admin.php?page=wpo_wcpdf_options_page&tab=ubl' ) ) . '" target="_blank">' . __( 'UBL Tab', 'woocommerce-pdf-invoices-packing-slips' ) . '</a>'
```
Perhaps rename this to `UBL settings` so it is the same as the `General settings` link.

Terminator-Barbapapa reviewed (approved):
I think we are there. Awesome work! 🎊
